pjlib/include/pj/argparse.h

```diff
@@ -0,0 +1,198 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+#ifndef __PJ_ARGPARSE_H__
+#define __PJ_ARGPARSE_H__
+
+/**
+ * @file argparse.h
+ * @brief Command line argument parser
+ */
+#include <pj/ctype.h>
+#include <pj/errno.h>
+#include <pj/string.h>
+
+PJ_BEGIN_DECL
+
+/**
+ * Define function to display parsing error.
+ */
+#ifndef PJ_ARGPARSE_ERROR
+#  include <stdio.h>
+#  define PJ_ARGPARSE_ERROR(fmt, arg) printf(fmt "\n", arg)
+#endif
+
+
+/**
+ * @defgroup PJ_ARGPARSE Command line argument parser
+ * @ingroup PJ_MISC
+ * @{
+ *
+ * This module provides header only utilities to parse command line arguments.
+ * This is mostly used by PJSIP test and sample apps. Note that there is
+ * getopt() implementation in PJLIB-UTIL (but it's in PJLIB-UTIL, so it can't
+ * be used by PJLIB)
+ */
+
+/**
+ * Peek the next possible option from argv. An argument is considered an
+ * option if it starts with "-" and followed by at least another letter that
+ * is not digit or starts with "--" and followed by a letter.
+ *
+ * @param argv      The argv, which must be null terminated.
+ *
+ * @return next option or NULL.
+ */
+PJ_INLINE(char*) pj_argparse_peek_next_option(char *const argv[])
+{
+    while (*argv) {
+        const char *arg = *argv;
+        if ((*arg=='-' && *(arg+1) && !pj_isdigit(*(arg+1))) ||
+            (*arg=='-' && *(arg+1)=='-' && *(arg+2)))
+        {
+            return *argv;
+        }
+        ++argv;
+    }
+    return NULL;
+}
+
+/**
+ * Check that an option exists, without modifying argv.
+ *
+ * @param opt       The option to find, e.g. "-h", "--help"
+ * @param argv      The argv, which must be null terminated.
+ *
+ * @return PJ_TRUE if the option exists, else PJ_FALSE.
+ */
+PJ_INLINE(pj_bool_t) pj_argparse_exists(const char *opt, char *const argv[])
+{
+    int i;
+    for (i=1; argv[i]; ++i) {
+        if (pj_ansi_strcmp(argv[i], opt)==0)
+            return PJ_TRUE;
+    }
+    return PJ_FALSE;
+}
+
+/**
+ * Check for an option and if it exists, returns PJ_TRUE remove that option
+ * from argc/argv.
+ *
+ * @param opt       The option to find, e.g. "-h", "--help"
+ * @param argc      Pointer to argc.
+ * @param argv      Null terminated argv.
+ *
+ * @return PJ_TRUE if the option exists, else PJ_FALSE.
+ */
+PJ_INLINE(pj_bool_t) pj_argparse_get_bool(const char *opt, int *argc, char *argv[])
+{
+    int i;
+    for (i=1; argv[i]; ++i) {
+        if (pj_ansi_strcmp(argv[i], opt)==0) {
+            pj_memmove(&argv[i], &argv[i+1], ((*argc)-i)*sizeof(char*));
+            (*argc)--;
+            return PJ_TRUE;
+        }
+    }
+    return PJ_FALSE;
+}
+
+/**
+ * Check for an option and if it exists, get the value and remove both
+ * the option the the value from argc/argv. Note that the function only
+ * supports whitespace as separator between option and value (i.e. equal
```

Comment:
a single whitespace or whitespaces?

Reply:
Added "limitations" section in this PR and in argparse doxygen to clarify the problem. Also doxygen comments updated to use "whitespace(s)".